packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,562 @@
+import type {
+  AccountsControllerAccountAddedEvent,
+  AccountsControllerAccountRemovedEvent,
+  AccountsControllerListMultichainAccountsAction,
+} from '@metamask/accounts-controller';
+import {
+  BaseController,
+  type ControllerGetStateAction,
+  type ControllerStateChangeEvent,
+  type RestrictedControllerMessenger,
+} from '@metamask/base-controller';
+import { isEvmAccountType } from '@metamask/keyring-api';
+import type {
+  AccountAssetListUpdatedEventPayload,
+  CaipAssetType,
+  CaipAssetTypeOrId,
+} from '@metamask/keyring-api';
+import type { InternalAccount } from '@metamask/keyring-internal-api';
+import { KeyringClient } from '@metamask/keyring-snap-client';
+import type {
+  GetPermissions,
+  PermissionConstraint,
+  SubjectPermissions,
+} from '@metamask/permission-controller';
+import type {
+  GetAllSnaps,
+  HandleSnapRequest,
+} from '@metamask/snaps-controllers';
+import type { FungibleAssetMetadata, Snap, SnapId } from '@metamask/snaps-sdk';
+import { HandlerType } from '@metamask/snaps-utils';
+import {
+  hasProperty,
+  isCaipAssetType,
+  parseCaipAssetType,
+  type CaipChainId,
+} from '@metamask/utils';
+import type { Json, JsonRpcRequest } from '@metamask/utils';
+import type { MutexInterface } from 'async-mutex';
+import { Mutex } from 'async-mutex';
+import { v4 as uuid } from 'uuid';
+
+import { getChainIdsCaveat } from './utils';
+import type { AccountsControllerAccountAssetListUpdatedEvent } from '../../../accounts-controller/src/AccountsController';
+
+const controllerName = 'MultichainAssetsController';
+
+export type MultichainAssetsControllerState = {
+  metadata: {
+    [asset: CaipAssetType]: FungibleAssetMetadata;
+  };
+  allNonEvmTokens: { [account: string]: CaipAssetType[] };
+};
```

Comment:
What's the benefits of having the two keys and not a single key that combines both ? Like: 
```ts
assets: {
  [account: string]: FungibleAssetMetadata
}
```

This would make it easier when we have to display this data, especially that there is no case where we will only use the CAIP asset ID without the metadata.

Reply:
Hi @GuillaumeRx 👋  Sorry; Are you asking why we are having both metadata and allNonEvmTokens instead of just assets?

```suggestion
assets: {
  [account: string]: FungibleAssetMetadata
}
```

Reply:
As discussed in DMs the concerns to keep the two state keys is valid. Closing.

Reply:
Resolving this after our discussion 🙏